src/server.cpp

```diff
@@ -2722,6 +2767,82 @@ void Server::UpdateCrafting(Player* player)
 	plist->changeItem(0, preview);
 }
 
+bool Server::handleChat(const std::string &name, const std::wstring &wname,
+	const std::wstring &wmessage, std::wstring &answer_to_sender,
+	u16 peer_id_to_avoid_sending)
+{
+	// If something goes wrong, this player is to blame
+	RollbackScopeActor rollback_scope(m_rollback,
+		std::string("player:") + name);
+
+	// Line to send
+	std::wstring line;
+	// Whether to send to the player that sent the line, or to all players
+	bool send_to_sender_only = false;
+
+	// Run script hook
+	bool ate = m_script->on_chat_message(name,
+		wide_to_utf8(wmessage));
+	// If script ate the message, don't proceed
+	if (ate)
+		return false;
+
+	// Commands are implemented in Lua, so only catch invalid
+	// commands that were not "eaten" and send an error back
+	if (wmessage[0] == L'/') {
+		std::wstring wcmd = wmessage.substr(1);
+		send_to_sender_only = true;
+		if (wcmd.length() == 0)
+			line += L"-!- Empty command";
+		else
+			line += L"-!- Invalid command: " + str_split(wcmd, L' ')[0];
+	} else {
+		line += L"<";
+		line += wname;
+		line += L"> ";
+		line += wmessage;
+	}
+
+
+	if (line != L"") {
+		/*
+			Tell calling method to send the message to sender
+		*/
+		if (send_to_sender_only) {
+			answer_to_sender = line;
+			return true;
+		} else {
+			/*
+				Send the message to others
+			*/
+			actionstream << "CHAT: " << wide_to_narrow(line) << std::endl;
+
+			std::vector<u16> clients = m_clients.getClientIDs();
+
+			for (std::vector<u16>::iterator i = clients.begin();
+				i != clients.end(); ++i) {
+					if (*i != peer_id_to_avoid_sending)
```

Comment:
I think your indentation is screwed up here...
Also wouldn't it be shorter to use array notation?